### PR TITLE
salus: added save/restore for vector context on context switch

### PR DESCRIPTION
**Pull request by stillson**
-added support for compiler target with vector extension
-various additions for vector registers
-runtime checking that we can support register width
-test code to check register restor and save
-extended some error messages in tellus

### CODE REVIEW

File: drivers/src/cpu.rs [hunk 1]
```diff
@@ -79,6 +81,16 @@ fn isa_string_has_extension(isa_string: &str, extension: &str) -> bool {
     isa_string.split('_').any(|f| f == extension)
 }
 
+fn isa_string_has_base_extension(isa_string: &str, extension: char) -> bool {
+    isa_string
+        .split('_')
+        .next()
+        .expect("ISA string too short")
```
Comment (atulkharerivos) [inline drivers/src/cpu.rs:88]: Should we panic if no features are detected?

File: riscv-regs/src/regs.rs [hunk 2]
```diff
@@ -117,3 +117,22 @@ impl GeneralPurposeRegisters {
 #[derive(Default)]
 #[repr(C)]
 pub struct FloatingPointRegisters([u64; 32]);
+
+/// The vector register file. We don't expect to directly interact with a guest's vector state
+/// other than for saving/restoring the registers, so simply treat the register file as an array
+/// of 256b values. This actually depends on the vlenb csr, so if the register is greater than 256
+/// bits (i.e. the vlenb csr is greater than 32) we will need to increase this.
+
+// the width of a vector register in bytes
```
Comment (atulkharerivos) [inline riscv-regs/src/regs.rs:126]: Nit: s/the/The/

File: src/main.rs [hunk 3]
```diff
@@ -402,6 +404,28 @@ pub fn setup_csrs() {
     trap::install_trap_handler();
 }
 
+#[cfg(target_feature = "v")]
+fn verify_vector_registers() {
+    // Because we just ran setup_csrs(), we know vectors are off
+    // Turn vectors on
+    CSR.sstatus.read_and_set_bits(0x200);
```
Comment (atulkharerivos) [inline src/main.rs:411]: Nit: Consider defining a const value for 0x200.

File: src/vectors.S [hunk 4]
```diff
@@ -0,0 +1,112 @@
+.section .text.init
+.global _run_guest_vector,_guest_exit_vector
+
+// t5 -> cached sstatus
+// restore vectors for guest
```
Comment (atulkharerivos) [inline src/vectors.S:5]: Nit (here and elsewhere): Capitalize first letter of sentence if applicable.

File: drivers/src/cpu.rs (hunk 1 above)
Comment (stillson) [reply drivers/src/cpu.rs:88]: is that even possible? next() will just return rv64i....

there is always at least one, isn't there?

File: riscv-regs/src/regs.rs (hunk 2 above)
Comment (stillson) [reply riscv-regs/src/regs.rs:126]: fixed

File: src/main.rs (hunk 3 above)
Comment (stillson) [reply src/main.rs:411]: There should be a way of getting from the tock register definition, but I couldn't figure it out (in 2 minutes) and I knew 0x200 off the top of my head. 

Will Fix

File: drivers/src/cpu.rs (hunk 1 above)
Comment (atulkharerivos) [reply drivers/src/cpu.rs:88]: Sure, it's nearly certain that we'll have at least one feature, but what I meant is that we probably don't need to go out of our way to panic (when we can just easily mark the feature as not detected).

File: src/vectors.S (hunk 4 above)
Comment (stillson) [reply src/vectors.S:5]: fixed

**Pull request closed without merging**